### ISSUE TITLE
typescript.tsdk "must be a user setting"

### ISSUE THREAD

**Issue by kylecordes** [Microsoft/vscode#19451]
- VSCode Version: Version 1.9.0-insider (1.9.0-insider) 6ba4282fd27bacacbf8482472a0519560720fb9c 2017-01-26T07:24:06.051Z

- OS Version: OSX 10.12.3

Steps to Reproduce:

1. Open any project that's been working fine for a long time with today's new insiders build
2. Observe error

As of today's new insiders build, VSCode  no longer accepts the cut the following very common settings:

```
  "typescript.check.workspaceVersion": false,
  "typescript.tsdk": "node_modules/typescript/lib"
```

As documented:

https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions

So this is both a bug report and a question. The question is how to achieve the following. The bug report is that there does not appear to be a way to achieve it.

Many of us work on many different projects with considerable version skew of the underlying libraries and tooling. This is a per project fact, not a per developer fact. There needs to be away to specify:

* Inside a project
* in the source code of the project, not in some other settings that a developer must learn about and then go adjust somewhere
* ideally, right here in the settings file, seems like the most reasonable place
* The fact that for this project it is desirable to use the same TypeScript version for IDE support as is used in the project itself.

There seems to be a new behavior in which as a user I am asked to click in the status bar to try to ask Code to use a certain version. That is a slightly nice developer experience, but it misses the point of the above.

Is there a new way to specify this thing that the old settings specified? (Or better yet, is there a way to make it just be the default that always happens all the time?  Offhand it seems that using a **different** version of TypeScript for IDE support than a project uses to compile, should be the unusual case requiring a special setting and jumping through hoops.)

**Comment by kylecordes** [comment]
One more thought on this. Thinking through the security implications, it is clear why removing the old behavior (where the IDE would merrily run any file in the project) is wise. I doubt anyone would object to that. The question is how to keep the "use the matching version of TypeScript" ability. The most ideal behavior would be for Code to see the old setting, recognize that its purpose was the TypeScript version match, and offer to replace it with a new setting which more explicitly means TypeScript version match, and is implemented in a way that presents a much smaller security risk surface.

**Comment by mjbvz** [comment]
//CC @kieferrm 

Thanks for the feedback. Here's some documentation on how the new flow is currently supposed to work: https://github.com/Microsoft/vscode-docs/pull/811 The new flow still allows you to switch to the workspace version of TS by clicking on the TS version number in the status bar.

We're still tweaking and tuning this new flow, so hearing your thoughts is very helpful and we'll certainly take them into account while making improvements. I'll also look into seeing if we can improve the upgrade experience for existing workspaces

**Comment by rob3c** [comment]
I think anything that requires an explicit action by a dev is a bad idea. Please give us the ability - or rather, please give us *back* the ability - to automate this on a per-workspace basis for a "pit of success" type experience for developers.

**Comment by kylecordes** [comment]
@mjbvz The new easy click is great for adoption. It demos well and lets people understand the most critical first minutes that they can set the version they need. Easing adoption is super important, it is essentially the metric of long-term project success.

But... Clicking a button to change the version is scarcely even a feature, from the point of view of an ongoing project past that initial adoption.  An ideal implementation would keep the button to help adoption, but have a long-term per-project plain-text settings-file setting instead of the button thereafter.

**Comment by joelday** [comment]
I want to be able to point to `built/local` in a clone of typescript without jumping through a bunch of hoops or changing my user-level settings.

**Comment by joelday** [comment]
Here's another fundamental problem: If I change the user level tsdk path to my local TypeScript build, I'm now using that local, probably broken build of TypeScript on *itself*.

**Comment by joelday** [comment]
Yeah, sorry, as soon as the January release is out, my workflow is completely broken and I have no idea how to work around it.

**Comment by mjbvz** [comment]
@kylecordes Thanks. I'm working right now on improving the 1.8 to 1.9 upgrade experience for workspaces with a `typescript.tsdk` setting or workspaces with a local version of TS installed. The current plan is to show a message when the user first loads these sort of workspaces. This message would ask users to select either the local or the workspace version of TS.

I think we've decided to move away from using the `typescript.tsdk` setting in a workspace though. For the majority of TypeScript 2.x users, this should not be a problem as the `typescript.tsdk` setting is only used for intellisense and other editor features, not for compiling the TypeScript code itself. The bundled TS 2.1.5 will just work on TS 2.0.x and TS 2.1.x code.

The two cases where using the bundled TS version may not be appropriate:

* Your project uses TS 1.x. There were  some breaking language changes between TS 1.x and TS 2.x
* You want to use the latest language features for TS, such as using `typescript@next`

In those cases, we support using the version of typescript installed into a workspace using `npm install typescript`. When you first load a workspace that has typescript installed, you will be asked if you want to use the workspace's version of TypeScript or use the default VSCode version of TypeScript. However, with this new flow, there is no longer be a workspace setting that explicitly defines which version of TS to use.

---

@rob3c Executing code from a user workspace is a security risk, so we need some sort of user action here. We're trying to make this as painless and clear as possible but, no matter what, you'll still have to take an action when you first load a workspace if you want to use a local executable

--- 

@joelday Thanks for the feedback.

We do want help extensions also avoid this issue, but the trusted workspaces flow that was prototyped in insiders was not ideal. Using it, extensions still had to implement custom logic for reading possibly unsafe settings value so, after discussion, we decided to scale back what VSCode provides and focus on creating getting good experiences for each extension. `typescript.tsdk` is probably the most commonly used unsafe workspace setting of this sort.

As for supporting workspace installs of TypeScript in different directories, we'll think about how this can be supported safely.

//cc @kieferrm

**Comment by kylecordes** [comment]
@mjbvz I'm a little disappointed but it sounds like the things several of us are looking for has already been decided against. Here will ask about a couple of specific scenarios and see if there is any possibility of them being addressed, perhaps by some other means?

> Your project uses TS 1.x. There were some breaking language changes between TS 1.x and TS 2.x

Yes, we have a couple of those. Is it possible that even if a fine-grained TypeScript version has been decided out of scope, that a project settings file could at least capture "this is a TypeScript 1.x project" versus "this is a TypeScript 2.0.x project" etc.?

> You want to use the latest language features for TS, such as using typescript@next

We have a couple of those also. Any hope?  Other than putting something in our README file and directing developers to always read the README file to find out if they need to click the right button in the UI?

> When you first load a workspace that has typescript installed

100% of our projects and therefore workspaces have TypeScript installed in the package.  I had assumed this was the typical case, there might be data showing that I am very wrong of course. Assuming that we can't have the feature discussed earlier, could there at least be a feature to make Code not prompt for every one of our projects the first time it is opened?  A way of answering the question once, globally?

I'm still concerned that the mechanism as described will inevitably lead to version skew, as some of our developers answer the question one way and others answer it the other way.  Are there any other features being considered that could manage version skew?  I mean automatically, not by an oral history of the team about which choice you are supposed to make when prompted by the blue bar at the top for each project.

It's possible that the use cases that I and several other people here asking about, really are oddballs, and that the new solution is best for most developers. The normal answer for that kind of case is an add-on. Do you happen to know if there are the right extension points available in the add-on mechanism, so that someone could create an add-on to capture this setting in a project-level text settings file? That is typically a best answer when a sub-population of users needs something out of scope for a base tool.

**Comment by rob3c** [comment]
@mjbvz Thanks for protecting us from ourselves! We're obviously incapable of using our own judgement to weigh the possible risks vs lost productivity. Fortunately, there isn't even a configurable option left as a later temptation...you know, in case we were to start thinking of ourselves as adults with the power to take responsibility for our own actions.

I wonder, though... The security risk seems to have been there all along, _and yet somehow the world didn't end_ while we were busy getting work done across projects and language versions. Hmm... maybe we just aren't downloading and blindly building enough random projects with rogue vscode workspace settings from the internet on our team.

At least now my coworkers and I won't be vulnerable once we've manually approved a local path in our local storage instead of that scary workplace settings file...even if someone were to push a rogue binary to that local repo path once they know we've all approved it...

**Comment by thorn0** [comment]
Trusted workspaces look like a good enough solution to this problem.

> the trusted workspaces flow that was prototyped in insiders was not ideal. Using it, extensions still had to implement custom logic for reading possibly unsafe settings value so

Why is it 'possibly unsafe'?

**Comment by kieferrm** [comment]
@thorn0 @joelday @kylecordes 

Any contribution to settings (also those from extensions) can now be marked as `isExecutable`. If marked, those settings can only be specified in User Settings. The trusted workspaces flow we had prototyped in insiders was a way to nevertheless allow workspaces to define such settings, but only after a user had explicitly approved a workspace to do so. The team rejected this general solution primarily for the complexity and potential misunderstandings around the notion of "trusting a workspace". A user does not only trust the workspace but all the repositories (i.e. remotes) the workspace comes from as well as all current and future contributors to that repository. 

The consequence of not having a general concept of trust is that we need a custom solution for each setting that is marked as `isExecutable`. What @mjbvz described is the flow we came up with for `typescript.tsdk`. Over the last couple of days, we tweak and polished the flow. We now prompt the user when a workspace has the `tsdk` setting and is opened for the first time and we ask whether the workspace's TS version or the VS Code global one should be used. (The VS Code global one can still be determined by setting the `tsdk` setting in the User Settings.) Once the user chooses the workspace version, all future changes to the workspace TS version are picked up. Factually, by choosing the workspace version, the user now marks the workspace as trusted in respect to whatever will happen in the workspaces TS version. This however is mitigated by the fact that most TS versions are not part of the repository but are fetched from npm when running `npm install`.

BTW, we are working with the TS team to find ways how we can automatically verify a version of TS as safe. For example, but signing the TS versions we submit to npm.

@kylecordes it is possible to create an extension that, when it finds the contents of the workspace to match a certain condition, will set the marker that the workspace version of TS should be used. Users need to take explicit steps to install an extension, so they should be aware of what they expose themselves to.

**Comment by joelday** [comment]
@kieferrm Even though it's a niche audience, I'd like to know how you're supposed to develop/test and test a language and/or language extension using that local build that uses an isExecutable setting to reference it.

I like to have a workspace that uses my local build of TypeScript so that I can interactively test my changes to tsserver. I *need to know how to do that now.*

**Comment by kylecordes** [comment]
@kieferrm Oh yes, the security concerns are highly justified. I'm very eager to see VSCode not create something akin to the expectations developed widely around NPM packages, which is that upon install they can execute arbitrary code under the identity of the developers user (!!). I think anyone who shares any security concerns should fully endorse these efforts to nail down the security better. Thank you to the whole team.

That's what I was trying to limit my concerns I've been writing about, specifically to specifying the TypeScript versions. Something like this:

* I am running VSCode, from Microsoft
* It has a TypeScript add-on shipped with it in the box, from Microsoft
* My project has a package file, it specifies a version of TypeScript. As far as I know, I should be able to trust that any of the versions of TypeScript that Microsoft is shipped, should be safe to run. Or if some are unsafe, the add-on should be able to blacklist those versions

The only thing missing is for the built-in Microsoft TypeScript support running in Microsoft VSCode, sometimes even running on Microsoft Windows, to go look at a text file and load up the correct version of Microsoft TypeScript. Of course the purpose of this repetition is to point out that I obviously already trust Microsoft, there is no further security risk added by specifying which TypeScript version to execute and then executing it.

Of course you're getting right to the crux of the matter with your comment about signing. Something like that seems possible, but a lot of scope, would need a lot of buy-in across the NPM ecosystem to work in a general way. For this local specific case though, VSCode could retrieve the necessary version of TypeScript itself, and execute it. This still might warrant some simple local implementation of signing, but it would not need any broader buy-in. (Or most simply, VSCode could ship with the latest 1.x, 2.0.x, and 2.1.x versions of TypeScript and merely select among those: no download or signing needed at all. Ugly, but a viable bridge to a later solution.)

Everything I wrote above is about TypeScript, but I think the same applies for other add-ons and other language support.

Thanks for the tip on the extension. I'm confident something like that will emerge if we can't get project-setting-level TypeScript version selection back into the base product. I'm going to sit tight for a while and hope the desirable capability emerges before digging in myself.

**Comment by joelday** [comment]
So, I could publish an extension that adds a command to:
- Get the unique id for the current workspace.
- Copy the current user data directory to a new temporary user data directory.
- Forces a new window to open that uses that directory.
- The extension then dynamically merges the workspace settings file onto a copy of the original user settings file.
- The extension watches the base directory in order to synchronize changes.

**Comment by joelday** [comment]
An additional consideration: In Visual Studio, I can point the language service to whatever I want by overriding the MSBuild property. Are they going to disable that, too?

**Comment by rotemdan** [comment]
I've read this discussion in detail and, too, is concerned about the loss of control introduced here. I appreciate the intention and consideration given to feedback. I think there are several things that need to be noted, though:

* `npm install` has the [`preinstall` and `postinstall` script hooks](https://docs.npmjs.com/misc/scripts) that can run any arbitrary command specified, both before and after the package has been installed. This can certainly be used to execute malicious code. Same goes for `npm test`, `npm <any command>`, `gulp`, `grunt` etc. In this light, I'm not sure if relying on an installation of TS that was acquired from the `npm` registry would provide a significant practical increase in safety.

* Some code-bases would not even compile without a specific version of the TS compiler. For example, I have several projects that must use the `next` (nightly build) version of TypeScript, otherwise there would be errors both at compile-time and run-time.

Considering this I'll try to give an (illustrative) suggestion:

* Since many TS workspaces include a `package.json` file in the root directory. VS Code could default to the version of `typescript` specified there (otherwise fall-back to the user setting). If a true measure of security is desired, I guess it might even be necessary to independently install that particular version by itself in a separate directory, which might even be protected under administrator permissions.

Of course some of this sounds like an overkill, because it is. And since this requires `package.json` file to already exist in the workspace directory then it is very likely that it would be used with `npm install`, counteracting most benefits from such strict safeguarding of the TS compiler.

I can also understand the need to allow for using custom builds of the compiler itself. Unfortunately it doesn't seem like that can be specified through a version number or a tag. I'm not sure exactly how that can be expressed without a particular path. For what it's worth it seems acceptable to limit the workspace-level `tsdk` path to be contained somewhere within the workspace directory tree (if needed, a symbolic/hard-link could be manually set-up to point to the actual path).

**Comment by kieferrm** [comment]
@joelday If the `tsdk` user setting is not an option then the following should work for you for now (assuming I understand your setup correctly):  In the workspace that should use the TS version under development, you symlink `<workspaceroot>/node_modules/typescript/lib` to the TS you want to use. You then tell VS to use the workspace version of TS for this workspace.  If you are using `git`, the symlink is no issue and you can have it in your repo. We are discussing folder-specific user settings for the next iteraton (like what we have done for language-specific settings in this iteration). With that the need for symlinking would vanish.

**Comment by joelday** [comment]
@kieferrm That works. Thank you!

**Comment by kieferrm** [comment]
@rotemdan Agreed, the `preinstall` and `post install` hooks are an issue. My point here was only that if the user runs `npm install typescript` then this should be considered reasonably safe. `npm install`, however, in a folder that has a `package.json` that you never looked at and comes from a source you don't know is a very different animal.

@kylecordes I think we could introduce signing that would be specific to the TS npm package without impacting the broader npm ecosystem.

@kylecordes @rotemdan Both of you propose a similar approach, i.e. that VS Code manages "trusted" Typescript versions which are used for providing language features and are picked based on a version setting (that also could be specified at the workspace level) or a combination of `package.json` and a setting. We discussed that as well. I created #19567 to make sure discussions from here on are accessible.

**Comment by rotemdan** [comment]
@kieferrm 

I appreciate the effort that's being put in trying to ensure consistent security guarantees for scripts and processes that are run directly by VS Code. I feel that's the main goal here, and that in practice it seems to serve mostly the editor itself, as the user's machine is still vulnerable in many other unrelated ways.

Creating an infrastructure for trusted execution of TS language services/compilers could be seen as a bit over-the-top for a general-purpose, language agnostic editor. It might eventually lead to the same to be done for TSLint (including its custom rules), JSLint, Go (and its tooling), Python (and its tooling), Roslyn/OmniSharp/C# etc. Personally I'm not sure if that would be the best use of the VS Code team's time but I appreciate the intention and dedication.

Unfortunately it doesn't personally make me feel any "safer". It just seems like something that would introduce unnecessary difficulty and frustration without actually providing a lot of practical value. I liked the idea of having a self-contained workspace, that internally bundles all the build-tools, compilers, language services, linters it needs at one place, with the exact versions required.

Now this would mean I have an additional step to negotiate with VS Code every time I want to upgrade the TS compiler to the latest nightly build, or want to downgrade to the one from several days ago because of a new bug introduced etc. or use a custom build I've made with a new experimental feature. I trust this issues would be sorted out eventually, but to me this doesn't really feel like a highly fruitful endeavor.

**Comment by kieferrm** [comment]
In preparation of 1.9 we made a few additional tweaks to the flow. The most notable one is that we re-established the `tsdk` workspace setting but only consider it for execution after the user explicitly agreed to do so. This simplifies @joelday's flow. Everything remains the same.

**Comment by joelday** [comment]
@kieferrm Yay! Thank you!

**Comment by rotemdan** [comment]
@kieferrm Thank you for reconsidering this. I apologize if I ended sounding a bit harsh.

I considered suggesting a global, user-level setting:
```
typescript.allowWorkspaceTsdkPaths
```
That could be enabled after the user explicitly confirmed they understand the risks.

Or a per-workspace hidden setting that would be internally stored by VSCode after the user has agreed to use a custom TSDK for that particular workspace. Since this requires a secondary "hidden" configuration for each workspace, that is stored separately from its actual location, I wasn't sure if that was a good idea.

**Comment by rob3c** [comment]
@kieferrm Are you ~annoying~ prompting the user early enough ~with this security theater~ so that accepting the setting doesn't require a restart like using the statusbar to change typescript versions does?

**Comment by kieferrm** [comment]
@rotemdan The solution in place is a flag on a per workspace basis.

**Comment by kieferrm** [comment]
@rob3c if it wouldn't be early enough, what would be the point of it all?

**Comment by rotemdan** [comment]
@kieferrm I'm wondering where would the flag be stored? I'm guessing in one of VSCode's internal storage folders?

Something like:
```
%AppData%\Roaming\Code\User\workspaceStorage
```
?

**Comment by rob3c** [comment]
> @rob3c if it wouldn't be early enough, what would be the point of it all?

@kieferrm There's no point to the interruption either way IMO, but thanks for confirming that there won't be the additional delay.

**Comment by thorn0** [comment]
@kieferrm Will it store some checksum of the tsdk code and re-ask the user when this sum changes?

**Comment by kieferrm** [comment]
@rotemdan We use the browser's local storage.

**Comment by kieferrm** [comment]
@thorn0 not in 1.9.

**Comment by rob3c** [comment]
So, inconveniencing users by removing the tsdk setting was presumably a security decision to avoid arbitrary code execution, but enabling extension autoupdate via settings - especially enabling it by default, as done in the latest update - is considered acceptable? I'm confused...
